nautobot/ui/src/components/Navbar.js

```diff
@@ -22,25 +28,83 @@ import {
     PlatformIcon,
     SearchIcon,
     SecurityIcon,
+    Tooltip,
 } from "@nautobot/nautobot-ui";
 
 import RouterButton from "@components/RouterButton";
+import { FILTER_RESET_QUERY_PARAMS } from "./FiltersPanel";
 import { useGetNewUIReadyRoutesQuery } from "@utils/api";
 import { isRouteNewUIReady } from "@utils/navigation";
 import {
     currentUserSelector,
     isLoggedInSelector,
     getCurrentContextSelector,
 } from "@utils/store";
-import { useEffect } from "react";
 
 export function Navbar() {
     const isLoggedIn = useSelector(isLoggedInSelector);
     const currentContext = useSelector(getCurrentContextSelector);
     const currentUser = useSelector(currentUserSelector);
+    const { app_label, model_name, object_id } = useParams(); // For use in determining if user is on a list or detail view, or neither
     const location = useLocation();
+    const navigate = useNavigate(); // For use in navigating to the list view with the search params in the URL
     const { data: readyRoutes } = useGetNewUIReadyRoutesQuery();
 
+    const [searchParams, setSearchParams] = useSearchParams();
+
+    const debouncedOnChangeSearchBox = useMemo(() => {
+        /**
+         * Debounce the search box change handler to prevent excessive API calls.
+         *
+         * This will wait until the user has stopped typing for 300ms before
+         * calling the change handler.
+         *
+         * If the user is on a list view, the search params will be updated in the URL
+         * and the user will be navigated to the new URL.
+         *
+         * If the user is on a detail view, the user will be navigated to the list view
+         * with the search params in the URL.
+         */
+        const changeHandler = (event) => {
+            if (object_id) {
+                navigate(
+                    `/${app_label}/${model_name}/?q=${event.target.value}`
+                );
```

Comment:
It's safer to use `URLSearchParams` prototype to add query params, or at least format them with `encodeURIComponent` function. That way they are ensured to always be valid URL encoded strings, as opossed to raw input values, e.g. `&` vs. `%26`, etc.

```suggestion
                navigate(
                    `/${app_label}/${model_name}/?${new URLSearchParams([
                        ["q", event.target.value],
                    ])}`
                );
```